File: documentation-examples-scala/src/scala/dev/typr/foundations/docs/core/BatchOperations.scala
package dev.typr.foundations.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.Fragment.sql
import dev.typr.foundationssc.data.*

import java.sql.Connection
import java.time.Instant

@SuppressWarnings(Array("unused"))
object BatchOperations:
  case class Product(id: Int, name: String, price: BigDecimal, createdAt: Instant)

  val productCodec: RowCodecNamed[Product] = RowCodec.namedBuilder[Product]()
    .field("id", PgTypes.int4)(_.id)
    .field("name", PgTypes.text)(_.name)
    .field("price", PgTypes.numeric)(_.price)
    .field("created_at", PgTypes.timestamptz)(_.createdAt)
    .build(Product.apply)

  val conn: Connection = null // placeholder

  //start
  // Batch insert — all columns as parameters
  val insertAll: RowTemplate.Update[Product] =
    sql"INSERT INTO product ("
      .append(productCodec.columnList)
      .append(") VALUES (")
      .paramRow(productCodec)
      .append(")")
      .update()

  def insertProducts(products: List[Product]): Array[Int] =
    insertAll.onMany(products.iterator).run(conn)

  // Batch insert — skip auto-generated ID column
  val insertAutoId: RowTemplate.Update[Product] =
    sql"INSERT INTO product (name, price, created_at) VALUES ("
      .paramRow(productCodec, "id")
      .append(")")
      .update()

  def insertProductsAutoId(products: List[Product]): Array[Int] =
    insertAutoId.onMany(products.iterator).run(conn)
  //stop

File: documentation-examples-kotlin/src/kotlin/dev/typr/foundations/docs/core/BatchOperations.kt
package dev.typr.foundations.docs.core

import dev.typr.foundationskt.*
import java.math.BigDecimal
import java.sql.Connection
import java.time.Instant

@Suppress("unused")
class BatchOperations {
    data class Product(val id: Int, val name: String, val price: BigDecimal, val createdAt: Instant)

    val productCodec: RowCodecNamed<Product> =
        RowCodec.namedBuilder<Product>()
            .field("id", PgTypes.int4, Product::id)
            .field("name", PgTypes.text, Product::name)
            .field("price", PgTypes.numeric, Product::price)
            .field("created_at", PgTypes.timestamptz, Product::createdAt)
            .build(::Product)

    lateinit var conn: Connection

    //start
    // Batch insert — all columns as parameters
    val insertAll =
        sql { "INSERT INTO product (" }
            .append(productCodec.columnList)
            .append(") VALUES (")
            .paramRow(productCodec)
            .append(")")
            .update()

    fun insertProducts(products: List<Product>): IntArray =
        insertAll.onMany(products.iterator()).run(conn)

    // Batch insert — skip auto-generated ID column
    val insertAutoId =
        sql { "INSERT INTO product (name, price, created_at) VALUES (" }
            .paramRow(productCodec, "id")
            .append(")")
            .update()

    fun insertProductsAutoId(products: List<Product>): IntArray =
        insertAutoId.onMany(products.iterator()).run(conn)
    //stop
}
